pkg/activator/throttler.go

```diff
@@ -0,0 +1,173 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package activator
+
+import (
+	"github.com/knative/serving/pkg/queue"
+	"go.uber.org/zap"
+	"sync"
+	"github.com/knative/serving/pkg/client/informers/externalversions/serving/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"errors"
+	corev1 "k8s.io/api/core/v1"
+	"github.com/knative/serving/pkg/reconciler"
+	"fmt"
+)
+
+const capacityUpdateFailure = "updating capacity failed"
+
+// NewThrottler creates a new Throttler
+func NewThrottler(breakerParams queue.BreakerParams, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: breakerParams, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	breakerParams    queue.BreakerParams
+	logger           *zap.SugaredLogger
+	mux              sync.RWMutex
+}
+
+// Add potentially creates a new breaker for the revision
+// returns true if a new entry was added
+func (t *Throttler) AddIfAbsent(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = *breaker
+		return true
+	}
+	return false
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	var containerConcurrency int32
+	revision, exists, err := t.getRevision(rev.Namespace, rev.Name, t.revisionInformer)
+	if !exists {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	if err != nil {
+		return err
+	}
+	containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+
+	if containerConcurrency == 0 {
+		containerConcurrency = t.breakerParams.MaxConcurrency
+	}
+	if breaker, ok := t.breakers[rev]; ok {
+		err := breaker.UpdateConcurrency(size * containerConcurrency)
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("no breaker was found for the received revision: %s", rev)
+	}
+	return nil
+}
+
+// Try executes the `function` on the Breaker
+// Returns true if the function was executed, false otherwise
```

Comment:
The comment does not represent reality.
Please terminate sentences with a period (this is kind of general comment for the whole fileset).

Reply:
Updated